### ISSUE TITLE
MRA version of LowRank is not equivalent

### ISSUE THREAD

**Issue by dzilber** [katzfuss-group/GPvecchia#48]
A modified predictive process is equivalent to MRA under conditions r = (m,1) but the current code implementation shows a discrepancy between the two techniques.   Example code below demonstrates with plots where the results are similar but not equal.

I will push a version of the code that separates low rank via MPP ("firstm") and MRA in Vecchia_specify and allows the code below to show the difference.

```
source("../GPvecchia/server/importer.R")

#Generate 1D data
n = 100
locs=matrix(runif(n),ncol=1)
covparms=c(1,.2,1)
Om0 <- Matern(fields::rdist(locs),
              range=covparms[2],
              smoothness=covparms[3])
y=as.numeric(t(chol(Om0))%*%rnorm(n))+1
nuggets=rep(.1,n)
z = rnorm(n,y,sqrt(nuggets))

# Approximate posterior
m=5
# Low Rank
vecchia.approx.lr = vecchia_specify(locs,m,conditioning = "firstm")
preds.lr = vecchia_prediction(z,vecchia.approx.lr,covparms,nuggets)

# Equivalent MRA
vecchia.approx.mra=vecchia_specify(locs,m,conditioning = "mra",
                                   mra.options = list(r=c(m,1)) )
preds.mra =vecchia_prediction(z,vecchia.approx.mra,covparms,nuggets)

loc_ord=order(locs)


# Plot of predictions
par(mfrow=c(1,2))
plot(locs, z, main = "Predictions")
lines(locs[loc_ord], preds.lr$mu.obs[loc_ord], type="l", col=3)
lines(locs[loc_ord], preds.mra$mu.obs[loc_ord], type="l", col=2)
legend("topright", legend=c("z", "LR", "MRA"),col = c(1,3,2), 
       lty = c(NA,1,1), pch = c(1, NA, NA))

# Plot showing discrepancy
plot(locs[loc_ord],
     preds.mra$mu.obs[loc_ord]-preds.lr$mu.obs[loc_ord],
     type="p", col=2, main = "Difference, MRA-LR", pch=19)
abline(h=0)
```

[ticket48.pdf](https://github.com/katzfuss-group/GPvecchia/files/3143317/ticket48.pdf)

**Comment by marcinjurek** [comment]
Ok, I pushed some updates that now let MRA work without bugs when a prediction is desired. I need to look into this one. By the way, excellent job documenting the problem!

**Comment by marcinjurek** [comment]
So I found the problem. When I was developing the MRA code I just hardcoded only one covariance function (exponential with radius 1) that it calculates. In other words, if you set `conditioning = 'mra'` `covparams` are effectively ignored. This is not good and I'm aware of it. However, when you set `covparams=c(1, 1, 0.5)` in the code above, the discrepancy disappears. I will create a separate ticket for upgrading the code so that all Matern functions are calculated (we already have the C code for it, I just need to figure out how to hook it up). If it works for you with the `covparams` I mentioned, please close this ticket.

**Comment by marcinjurek** [comment]
By the way, I understand that the status quo severely limits the applicability of the mra code and that bypassing it when the `firstm` option is selected makes sense until #49 is resolved.

**Comment by marcinjurek** [comment]
I think this should be resolved in 25a7b6b. I tested it using the file you provided. Please test on your.
